base/bundle.hpp

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+#include <atomic>
+#include <experimental/optional>
+#include <functional>
+#include <list>
+#include <map>
+#include <memory>
+#include <mutex>
+
+#include "base/macros.hpp"
+#include "base/not_null.hpp"
+#include "base/status.hpp"
+
+namespace principia {
+namespace base {
+
+class Bundle {
+ public:
+  class Task;
+
+  Bundle(int threads);
+
+  // Returns the first non-OK status encountered.  All threads are joined;
+  // no calls to |Add| or |Join| may follow this call.
+  Status Join();
+
+  // The |Task| returned by |Add| may be used in calls to |Abort| or |JoinTask|;
+  // |this| shares ownership until |task| has returned or is dequeued by
+  // |Abort|.
+  std::shared_ptr<Task> Add(std::function<Status()> task);
```

Comment:
Calling the parameter `task` makes this unclear.  Here and in the comments below I would like to distinguish the function/callback from the `Task`.  (You use `body` elsewhere, that works better.)